// chat-callkit/lib/index.js
"use strict";

exports.__esModule = true;
exports["default"] = exports.CallkitContext = void 0;

var _react = _interopRequireDefault(require("react"));

require("./style.css");

require("./assets/icon/iconfont.css");

var _reactRedux = require("react-redux");

var _redux = _interopRequireDefault(require("./redux"));

var _layout = _interopRequireDefault(require("./layout"));

var _callManager = require("./modules/callManager");

var _message = require("./modules/message");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

window.callkit_store = _redux["default"];

var CallkitContext = _react["default"].createContext();

exports.CallkitContext = CallkitContext;

function Callkit(props) {
  return /*#__PURE__*/_react["default"].createElement(_reactRedux.Provider, {
    store: _redux["default"]
  }, /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(CallkitContext.Provider, {
    value: props
  }, /*#__PURE__*/_react["default"].createElement(_layout["default"], props))));
}

Callkit.init = function (appId, uid, conn) {
  if (typeof appId !== 'string') {
    throw new Error("invalid parameter appId: " + appId);
  } else if (typeof uid !== 'string') {
    throw new Error("invalid parameter uid: " + uid);
  } else if (typeof conn !== 'object') {
    throw new Error("invalid parameter conn: " + conn);
  }

  _callManager.callManager.init(appId, uid, conn);
};

Callkit.startCall = function (options) {
  if (typeof options.to !== 'string' && !(options.to instanceof Array)) {
    throw new Error("invalid parameter options.to: " + options.to);
  } else if (typeof options.channel !== 'string') {
    throw new Error("invalid parameter options.channel: " + options.channel);
  } else if (typeof options.accessToken !== 'string') {
    throw new Error("invalid parameter options.accessToken: " + options.accessToken);
  } else if (![0, 1, 2, 3].includes(options.callType)) {
    throw new Error("invalid parameter options.callType: " + options.callType);
  } else if (options.chatType === 'groupChat' && !options.groupId) {
    throw new Error("The groupId is required");
  }

  if (options.chatType === 'groupChat' && Array.isArray(options.to)) {
    var _callId = WebIM.conn.getUniqueId().toString();

    var channel = options.channel || Math.uuid(8);

    var _params = _extends({}, options, {
      callId: _callId,
      channel: channel
    });

    options.to.forEach(function (userId) {
      _params.to = userId;
      _params.chatType = 'singleChat';

      _callManager.callManager.startCall(_params);
    });
    (0, _message.sendTextMsg)(options.chatType, options.groupId, options.message, {
      action: 'invite',
      type: options.callType,
      msgType: 'rtcCallWithAgora',
      callId: _callId,
      channelName: channel,
      callerDevId: WebIM.conn.context.jid.clientResource,
      ts: Date.now()
    });
    return;
  }

  var callId = WebIM.conn.getUniqueId().toString();

  var params = _extends({}, options, {
    callId: callId
  });

  _callManager.callManager.startCall(params);

  return {
    callId: callId
  };
};

Callkit.answerCall = function (result, token) {
  if (typeof result !== 'boolean') {
    throw new Error("invalid parameter result: " + result);
  } else if (typeof token !== 'string') {
    throw new Error("invalid parameter token: " + token);
  }

  _callManager.callManager.answerCall(result, token);
};

Callkit.setUserIdMap = function (idMap) {
  if (typeof idMap !== 'object') {
    throw new Error("invalid parameter idMap: " + idMap);
  }

  _callManager.callManager.setUserIdMap(idMap);
};

var _default = Callkit;
exports["default"] = _default;
