// agora-chat-uikit/lib/EaseChat/thread/threadList/index.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _index = require("../../../EaseApp/index");

var _core = require("@material-ui/core");

var _thread = _interopRequireDefault(require("../../../redux/thread"));

var _index2 = require("../../../utils/index");

var _threadClose = _interopRequireDefault(require("../../../common/images/threadClose.png"));

require("./index.css");

var _avatar = _interopRequireDefault(require("../../../common/icons/avatar1.jpg"));

var _i18next = _interopRequireDefault(require("i18next"));

var _emoji = require("../../../common/emoji");

var _lodash = _interopRequireDefault(require("lodash"));

var _AppDB = _interopRequireDefault(require("../../../utils/AppDB"));

var _alert = require("../../../EaseChat/common/alert");

var _index4 = require("../../chat/index");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// import "../../../i18n";
var ThreadListPanel = function ThreadListPanel(_ref) {
  var anchorEl = _ref.anchorEl,
      onClose = _ref.onClose;
  var easeChatProps = (0, _react.useContext)(_index4.EaseChatContext);
  var onOpenThreadPanel = easeChatProps.onOpenThreadPanel;
  var dispatch = (0, _index.useDispatch)();
  var threadList = (0, _index.useSelector)(function (state) {
    var _state$thread;

    return (_state$thread = state.thread) === null || _state$thread === void 0 ? void 0 : _state$thread.threadList;
  }) || [];
  var curGroupRole = (0, _index.useSelector)(function (state) {
    var _state$thread2;

    return (_state$thread2 = state.thread) === null || _state$thread2 === void 0 ? void 0 : _state$thread2.curGroupRole;
  }) || '';

  var _useSelector = (0, _index.useSelector)(function (state) {
    return state.global.globalProps;
  }),
      chatType = _useSelector.chatType,
      to = _useSelector.to;

  var threadListDom = (0, _react.useRef)(null);
  (0, _react.useEffect)(function () {
    if (chatType === "groupChat" && anchorEl) {
      var dom = threadListDom.current;

      if (_reactDom["default"].findDOMNode(dom)) {
        dom.scrollTop = 0;
      }

      dispatch(_thread["default"].setThreadListEnd(false));
      var options = {
        groupId: to,
        limit: 20
      };
      dispatch(_thread["default"].getThreadsListOfGroup(options));
    }
  }, [anchorEl]);

  var renderMessage = function renderMessage(message) {
    switch (message.type) {
      case 'txt':
        return renderTxt(message.msg);

      case 'file':
        return "[" + _i18next["default"].t('File') + "]";

      case 'img':
        return "[" + _i18next["default"].t('Image') + "]";

      case 'audio':
        return "[" + _i18next["default"].t('Audio') + "]";

      case 'video':
        return "[" + _i18next["default"].t('Video') + "]";

      default:
        return '';
    }
  };

  var renderTxt = function renderTxt(txt) {
    if (txt === undefined) {
      return [];
    }

    var rnTxt = [];
    var match = null;
    var regex = /(\[.*?\])/g;
    var start = 0;
    var index = 0;

    while (match = regex.exec(txt)) {
      index = match.index;

      if (index > start) {
        rnTxt.push(txt.substring(start, index));
      }

      if (match[1] in _emoji.emoji.map) {
        var v = _emoji.emoji.map[match[1]];
        rnTxt.push( /*#__PURE__*/_react["default"].createElement("img", {
          key: v + Math.floor(Math.random() * 99 + 1),
          alt: v,
          src: require("../../../common/reactions/" + v)["default"],
          width: 20,
          height: 20
        }));
      } else {
        rnTxt.push(match[1]);
      }

      start = index + match[1].length;
    }

    rnTxt.push(txt.substring(start, txt.length));
    return rnTxt;
  };

  var _useState = (0, _react.useState)(false),
      isPullingDown = _useState[0],
      setIsPullingDown = _useState[1];

  var handleScroll = function handleScroll(e) {
    if (e.target.scrollHeight === e.target.scrollTop + e.target.clientHeight) {
      if (!isPullingDown) {
        setIsPullingDown(true);
        setTimeout(function () {
          setIsPullingDown(false);

          if (chatType === "groupChat") {
            var options = {
              groupId: to,
              limit: 20,
              isScroll: true
            };
            dispatch(_thread["default"].getThreadsListOfGroup(options));
          }
        }, 500);
      }
    }
  };

  var openThreadPanel = function openThreadPanel(option) {
    if (curGroupRole === 'member') {
      changeCurrentThreadInfo(option);
    } else {
      WebIM.conn.joinChatThread({
        chatThreadId: option.id
      }).then(function (res) {
        changeCurrentThreadInfo(option);
      })["catch"](function (e) {
        if (e.type === 1301) {
          changeCurrentThreadInfo(option);
        } else if (e.type === 1300) {
          _alert.message.warn(_i18next["default"].t('The thread has been disbanded'));
        }
      });
    }

    onOpenThreadPanel && onOpenThreadPanel(option);
  };

  var changeCurrentThreadInfo = function changeCurrentThreadInfo(option) {
    dispatch(_thread["default"].setCurrentThreadInfo(option)); //updatea setThreadOriginalMsg

    if (option.messageId) {
      _AppDB["default"].findLocalMessage('groupChat', option.messageId).then(function (res) {
        var msg = res.length === 1 ? res[0] : {};
        dispatch(_thread["default"].setThreadOriginalMsg(msg));
      });
    } //update the status of creatingThread


    dispatch(_thread["default"].setIsCreatingThread(false)); //open threadPanel

    dispatch(_thread["default"].updateThreadStates(true)); //close threadListPanel

    onClose();
  }; //The list is empty


  var renderDefaultList = function renderDefaultList() {
    if (threadList.length === 0) return /*#__PURE__*/_react["default"].createElement(_core.Box, {
      className: "tlp-default-tips"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "tlp-tips1"
    }, /*#__PURE__*/_react["default"].createElement("span", {
      className: "tlp-tips1-img"
    }), _i18next["default"].t('There are no Threads')), /*#__PURE__*/_react["default"].createElement("div", {
      className: "tlp-tips2"
    }, _i18next["default"].t('Create a Thread from a Group Chat Message')));
  };

  return /*#__PURE__*/_react["default"].createElement(_core.Popover, {
    open: Boolean(anchorEl),
    onClose: onClose,
    anchorEl: anchorEl,
    anchorOrigin: {
      vertical: "top",
      horizontal: "right"
    },
    transformOrigin: {
      vertical: "top",
      horizontal: "right"
    }
  }, /*#__PURE__*/_react["default"].createElement(_core.Box, {
    className: "threadListPanel"
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: "tlp-header-thread"
  }, /*#__PURE__*/_react["default"].createElement("span", {
    className: "tlp-header-title"
  }, _i18next["default"].t('Threads List')), /*#__PURE__*/_react["default"].createElement("div", {
    className: "tlp-header-icon",
    onClick: onClose
  }, /*#__PURE__*/_react["default"].createElement("img", {
    className: "tlp-header-icon-close",
    alt: "closeIcon",
    src: _threadClose["default"]
  }))), /*#__PURE__*/_react["default"].createElement("ul", {
    className: "tlp-list",
    ref: threadListDom,
    onScroll: handleScroll
  }, renderDefaultList(), threadList.length > 0 && threadList.map(function (option, index) {
    var _option$lastMessage;

    return /*#__PURE__*/_react["default"].createElement("li", {
      className: "tlp-item",
      key: index,
      onClick: function onClick(e) {
        return openThreadPanel(option);
      }
    }, /*#__PURE__*/_react["default"].createElement(_core.Box, {
      className: "tpl-item-con"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "tpl-item-name"
    }, option.name), /*#__PURE__*/_react["default"].createElement(_core.Box, {
      className: "tpl-item-bottom"
    }, /*#__PURE__*/_react["default"].createElement("img", {
      className: "tlp-avatar",
      alt: "Remy Sharp",
      src: _avatar["default"]
    }), /*#__PURE__*/_react["default"].createElement("span", {
      className: "tpl-item-owner"
    }, option.owner), option.lastMessage && /*#__PURE__*/_react["default"].createElement("span", {
      className: "tpl-item-msg"
    }, renderMessage(option.lastMessage)), /*#__PURE__*/_react["default"].createElement("span", {
      className: "tpl-item-time"
    }, (0, _index2.getTimeDiff)((_option$lastMessage = option.lastMessage) === null || _option$lastMessage === void 0 ? void 0 : _option$lastMessage.time)))));
  }))));
};

var _default = ThreadListPanel;
exports["default"] = _default;
module.exports = exports.default;
