// agora-chat-uikit/lib/EaseChat/thread/components/searchBox.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styles = require("@material-ui/styles");

var _searchgray = _interopRequireDefault(require("../../../common/images/searchgray.png"));

var _cancel = _interopRequireDefault(require("../../../common/images/cancel.png"));

var _i18next = _interopRequireDefault(require("i18next"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// import "../../../i18n";
var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    container: {
      height: "36px",
      width: '326px',
      display: "flex",
      justifyContent: 'space-between'
    },
    searchCon: {
      display: "flex",
      width: '253px',
      height: '36px',
      lineHeight: '36px',
      background: '#F2F2F2',
      borderRadius: '18px'
    },
    searchIcon: {
      height: '36px',
      width: '36px',
      cursor: 'pointer',
      background: "url(" + _searchgray["default"] + ") center center no-repeat"
    },
    searchInput: {
      height: '36px',
      lineHeight: '36px',
      width: '180px',
      border: 'none',
      background: 'none',
      outline: 'none',
      padding: '0',
      fontSize: '16px',
      color: '#000'
    },
    cancelIcon: {
      height: '36px',
      width: '36px',
      background: "url(" + _cancel["default"] + ") center center no-repeat",
      cursor: 'pointer'
    },
    cancel: {
      width: '95px',
      textAlign: 'center',
      height: '36px',
      lineHeight: '36px',
      fontWeight: '600',
      fontSize: '16px',
      color: '#005FFF',
      cursor: 'pointer'
    }
  };
});

var searchBox = function searchBox(props) {
  var classes = useStyles();

  var _useState = (0, _react.useState)(''),
      searchValue = _useState[0],
      setSearchValue = _useState[1];

  var changeSearchValue = function changeSearchValue(e) {
    setSearchValue(e.target.value);
  };

  var closeSearchBar = function closeSearchBar(state) {
    props.changeSearchBarState(state);
  };

  var search = function search() {
    props.searchThread(searchValue);
  };

  var clearInput = function clearInput(e) {
    setSearchValue('');
  };

  return /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.container
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.searchCon
  }, /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.searchIcon,
    onClick: search
  }), /*#__PURE__*/_react["default"].createElement("input", {
    className: classes.searchInput,
    value: searchValue,
    onChange: function onChange(e) {
      return changeSearchValue(e);
    }
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.cancelIcon,
    onClick: clearInput
  })), /*#__PURE__*/_react["default"].createElement("div", {
    className: classes.cancel,
    onClick: function onClick(e) {
      return closeSearchBar(false);
    }
  }, _i18next["default"].t("Cancel")));
};

var _default = searchBox;
exports["default"] = _default;
module.exports = exports.default;
