// agora-chat-uikit/lib/EaseChat/common/alert.js
"use strict";

exports.__esModule = true;
exports.Message = Message;
exports.message = exports.alert = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _lab = require("@material-ui/lab");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var wrap;
var message = {
  error: function error(content) {
    alert(content, 'error');
  },
  warn: function warn(content) {
    alert(content, 'warning');
  },
  info: function info(content) {
    alert(content, 'info');
  },
  success: function success(content) {
    alert(content, 'success');
  }
};
exports.message = message;

var alert = function alert(content, type) {
  if (type === void 0) {
    type = 'info';
  }

  if (!wrap) {
    wrap = document.createElement("div");
    wrap.style.cssText = "\n        line-height:1.5;\n        text-align:center;\n        color: #333;\n\t\tbox-sizing: border-box;\n\t\tmargin: 0;\n\t\tpadding: 0;\n\t\tlist-style: none;\n\t\tposition: fixed;\n\t\tz-index: 100000;\n\t\twidth: 100%;\n\t\ttop: 16px;\n\t\tleft: 0;\n\t\tpointer-events: none;";
  }

  if (wrap) {
    document.body && document.body.appendChild(wrap); //挂body上
  }

  var divs = document.createElement("div");
  wrap.appendChild(divs);

  _reactDom["default"].render( /*#__PURE__*/_react["default"].createElement(Message, {
    rootDom: wrap,
    parentDom: divs,
    content: content,
    type: type
  }), divs);
};

exports.alert = alert;

function Message(props) {
  var parentDom = props.parentDom,
      rootDom = props.rootDom,
      content = props.content,
      type = props.type;
  var unmount = (0, _react.useMemo)(function () {
    return function () {
      if (parentDom && rootDom) {
        _reactDom["default"].unmountComponentAtNode(parentDom);

        rootDom.removeChild(parentDom);
      }
    };
  }, [parentDom, rootDom]);
  (0, _react.useEffect)(function () {
    setTimeout(function () {
      unmount();
    }, 2000);
  }, [unmount]);
  return /*#__PURE__*/_react["default"].createElement(_lab.Alert, {
    severity: type
  }, content);
}
