// agora-chat-uikit/es/EaseChat/thread/components/searchBox.js
import React, { useState } from "react";
import { makeStyles } from "@material-ui/styles";
import searchgray from '../../../common/images/searchgray.png';
import searchCancel from '../../../common/images/cancel.png'; // import "../../../i18n";

import i18next from "i18next";
var useStyles = makeStyles(function (theme) {
  return {
    container: {
      height: "36px",
      width: '326px',
      display: "flex",
      justifyContent: 'space-between'
    },
    searchCon: {
      display: "flex",
      width: '253px',
      height: '36px',
      lineHeight: '36px',
      background: '#F2F2F2',
      borderRadius: '18px'
    },
    searchIcon: {
      height: '36px',
      width: '36px',
      cursor: 'pointer',
      background: "url(" + searchgray + ") center center no-repeat"
    },
    searchInput: {
      height: '36px',
      lineHeight: '36px',
      width: '180px',
      border: 'none',
      background: 'none',
      outline: 'none',
      padding: '0',
      fontSize: '16px',
      color: '#000'
    },
    cancelIcon: {
      height: '36px',
      width: '36px',
      background: "url(" + searchCancel + ") center center no-repeat",
      cursor: 'pointer'
    },
    cancel: {
      width: '95px',
      textAlign: 'center',
      height: '36px',
      lineHeight: '36px',
      fontWeight: '600',
      fontSize: '16px',
      color: '#005FFF',
      cursor: 'pointer'
    }
  };
});

var searchBox = function searchBox(props) {
  var classes = useStyles();

  var _useState = useState(''),
      searchValue = _useState[0],
      setSearchValue = _useState[1];

  var changeSearchValue = function changeSearchValue(e) {
    setSearchValue(e.target.value);
  };

  var closeSearchBar = function closeSearchBar(state) {
    props.changeSearchBarState(state);
  };

  var search = function search() {
    props.searchThread(searchValue);
  };

  var clearInput = function clearInput(e) {
    setSearchValue('');
  };

  return /*#__PURE__*/React.createElement("div", {
    className: classes.container
  }, /*#__PURE__*/React.createElement("div", {
    className: classes.searchCon
  }, /*#__PURE__*/React.createElement("div", {
    className: classes.searchIcon,
    onClick: search
  }), /*#__PURE__*/React.createElement("input", {
    className: classes.searchInput,
    value: searchValue,
    onChange: function onChange(e) {
      return changeSearchValue(e);
    }
  }), /*#__PURE__*/React.createElement("div", {
    className: classes.cancelIcon,
    onClick: clearInput
  })), /*#__PURE__*/React.createElement("div", {
    className: classes.cancel,
    onClick: function onClick(e) {
      return closeSearchBar(false);
    }
  }, i18next.t("Cancel")));
};

export default searchBox;
