// agora-chat-uikit/lib/redux/session.js
"use strict";

exports.__esModule = true;
exports.topSession = exports.setSessionList = exports.setJoinedGroups = exports.setCurrentSession = exports.sessionReducer = exports.pushSession = exports.deleteSession = exports["default"] = exports.INITIAL_STATE = void 0;

var _reduxsauce = require("reduxsauce");

var _seamlessImmutable = _interopRequireDefault(require("seamless-immutable"));

var _lodash = _interopRequireDefault(require("lodash"));

var _AppDB = _interopRequireDefault(require("../utils/AppDB"));

var _WebIM = _interopRequireDefault(require("../utils/WebIM"));

var _createReducer;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/* ------------- Types and Action Creators ------------- */
var _createActions = (0, _reduxsauce.createActions)({
  setSessionList: ['sessionList'],
  setCurrentSession: ['userId'],
  topSession: ['sessionId', 'sessionType', 'message'],
  deleteSession: ['sessionId'],
  pushSession: ['session'],
  setJoinedGroups: ["joinedGroups"],
  getSessionList: function getSessionList(userId) {
    return function (dispatch, getState) {
      _AppDB["default"].getSessionList(null, userId).then(function (res) {
        console.log('获取会话列表', res);
        dispatch(Creators.setSessionList(res));
      });
    };
  },
  getJoinedGroupList: function getJoinedGroupList() {
    return function (dispatch, getState) {
      _WebIM["default"].conn.getGroup().then(function (res) {
        console.log('res', res);
        var joinedGroups = res.data;
        dispatch(Creators.setJoinedGroups(joinedGroups));
      });
    };
  },
  _pushSession: function _pushSession(session) {
    return function (dispatch) {
      dispatch(Creators.pushSession(session));
    };
  }
}),
    Types = _createActions.Types,
    Creators = _createActions.Creators;

var _default = Creators;
exports["default"] = _default;
var INITIAL_STATE = (0, _seamlessImmutable["default"])({
  sessionList: [],
  currentSession: '',
  joinedGroups: []
});
/* ------------- Reducers ------------- */

exports.INITIAL_STATE = INITIAL_STATE;

var setSessionList = function setSessionList(state, _ref) {
  var sessionList = _ref.sessionList;
  var stateSession = state.sessionList ? state.sessionList : [];

  var concatSession = _lodash["default"].concat(stateSession, sessionList);

  var newConcatSession = _lodash["default"].uniqBy(concatSession, 'sessionId');

  return state.merge({
    sessionList: newConcatSession
  });
};

exports.setSessionList = setSessionList;

var setCurrentSession = function setCurrentSession(state, _ref2) {
  var userId = _ref2.userId;
  return state.merge({
    currentSession: userId
  });
};

exports.setCurrentSession = setCurrentSession;

var setJoinedGroups = function setJoinedGroups(state, _ref3) {
  var joinedGroups = _ref3.joinedGroups;
  return state.merge({
    joinedGroups: joinedGroups
  });
};

exports.setJoinedGroups = setJoinedGroups;

var topSession = function topSession(state, _ref4) {
  var sessionId = _ref4.sessionId,
      sessionType = _ref4.sessionType,
      message = _ref4.message;

  if ((message === null || message === void 0 ? void 0 : message.chatThread) && JSON.stringify(message.chatThread) !== '{}') {
    return state;
  }

  var sessionList = state.getIn(['sessionList'], (0, _seamlessImmutable["default"])([])).asMutable();
  var topSession = {
    sessionId: sessionId,
    sessionType: sessionType
  };
  sessionList.forEach(function (element, index) {
    if (sessionId === element.sessionId) {
      topSession = element;
      sessionList.splice(index, 1);
    }
  });
  sessionList.unshift(topSession);
  return state.merge({
    sessionList: sessionList
  });
};

exports.topSession = topSession;

var deleteSession = function deleteSession(state, _ref5) {
  var sessionId = _ref5.sessionId;
  var sessionList = state.sessionList.asMutable();
  sessionList = sessionList.filter(function (item) {
    return item.sessionId !== sessionId;
  });
  state = state.setIn(['sessionList'], sessionList);
  state = state.setIn(['currentSession', '']);
  return state;
};

exports.deleteSession = deleteSession;

var pushSession = function pushSession(state, _ref6) {
  var session = _ref6.session;
  var ary = state.sessionList.asMutable();

  var newSessionList = _lodash["default"].concat(ary, session);

  return state.setIn(['sessionList'], newSessionList);
};
/* ------------- Hookup Reducers To Types ------------- */


exports.pushSession = pushSession;
var sessionReducer = (0, _reduxsauce.createReducer)(INITIAL_STATE, (_createReducer = {}, _createReducer[Types.SET_SESSION_LIST] = setSessionList, _createReducer[Types.SET_CURRENT_SESSION] = setCurrentSession, _createReducer[Types.TOP_SESSION] = topSession, _createReducer[Types.DELETE_SESSION] = deleteSession, _createReducer[Types.PUSH_SESSION] = pushSession, _createReducer[Types.SET_JOINED_GROUPS] = setJoinedGroups, _createReducer));
exports.sessionReducer = sessionReducer;
