// chat-callkit/es/modules/message.js
import { WebIM, callManager } from './callManager';
import './utils';
import store from '../redux';
import { updateConfr, setCallStatus, CALLSTATUS } from '../redux/reducer';
var manager = {};
export var sendTextMsg = function sendTextMsg(chatType, to, message, ext) {
  var option = {
    chatType: chatType,
    type: "txt",
    to: to,
    msg: message,
    ext: ext
  };
  var msg = WebIM.message.create(option);
  WebIM.conn.send(msg);

  if (ext.type === 0 || ext.type === 1) {
    var _dispatch = store.dispatch;
    WebIM.rtc.timer = setTimeout(function () {
      console.warn('caller timeout');
      callManager.hangup('timeout', true);

      _dispatch(setCallStatus(CALLSTATUS.idle));
    }, 30000);
  }
};
export var sendCMDMsg = function sendCMDMsg(chatType, to, message, ext) {
  var id = WebIM.conn.getUniqueId();
  var msg = new WebIM.message('cmd', id);
  msg.set({
    to: to,
    action: 'rtcCall',
    ext: ext,
    success: function success(id, serverMsgId) {
      dispatch(setCallStatus(CALLSTATUS.alerting));
    },
    fail: function fail(e) {
      console.warn("callee send alert fail", msg);
    }
  });
  WebIM.conn.send(msg.body);
}; // callee

export var sendAlerting = function sendAlerting(to, calleeDevId, callId) {
  var getState = store.getState,
      dispatch = store.dispatch;
  var id = WebIM.conn.getUniqueId();
  var msg = new WebIM.message('cmd', id);

  if (!to) {
    return;
  }

  msg.set({
    to: to,
    action: 'rtcCall',
    ext: {
      action: 'alert',
      calleeDevId: WebIM.conn.context.jid.clientResource,
      callerDevId: calleeDevId,
      callId: callId,
      ts: Date.now(),
      msgType: 'rtcCallWithAgora'
    },
    success: function success(id, serverMsgId) {
      dispatch(setCallStatus(CALLSTATUS.alerting));
    },
    fail: function fail(e) {
      console.warn("callee send alert fail", msg);
    }
  });
  WebIM.conn.send(msg.body);
  WebIM.rtc.timer = setTimeout(function () {
    console.warn('callee timeout');
    callManager.hangup('timeout');
    dispatch(setCallStatus(CALLSTATUS.idle));
  }, 30000);
}; // caller

var confirmRing = function confirmRing(to, calleeDevId, callerDevId, callId) {
  var getState = store.getState,
      dispatch = store.dispatch;
  var confr = getState().confr;
  var currentCallId = confr.callId;
  var status = true;

  if (callId !== currentCallId) {
    console.warn('not current call', callId, currentCallId);
    status = false;
  }

  if (getState().callStatus > CALLSTATUS.receivedConfirmRing && ![2, 3].includes(confr.type)) {
    //in call
    console.warn('caller is busy', confr);
    status = false;
  } // if (confr.calleeDevId && confr.calleeDevId != calleeDevId){
  // 	console.warn('calleeDevId is different')
  // 	status = false
  // }


  if (callerDevId !== WebIM.conn.context.jid.clientResource) {
    console.warn('caller device is different');
    return;
  }

  var id = WebIM.conn.getUniqueId();
  var msg = new WebIM.message('cmd', id);
  msg.set({
    to: to,
    action: 'rtcCall',
    ext: {
      action: 'confirmRing',
      status: status,
      callerDevId: WebIM.conn.context.jid.clientResource,
      calleeDevId: calleeDevId,
      callId: callId,
      ts: Date.now(),
      msgType: 'rtcCallWithAgora'
    },
    success: function success(id, serverMsgId) {
      if (status) {
        if (getState().callStatus < CALLSTATUS.confirmRing) {
          dispatch(setCallStatus(CALLSTATUS.confirmRing));
        }
      }
    },
    fail: function fail(e) {
      console.warn("caller send confirmRing fail", msg);
    }
  });
  WebIM.conn.send(msg.body);
}; // callee


export var answerCall = function answerCall(result, info) {
  var getState = store.getState,
      dispatch = store.dispatch;
  info = info || {};
  var id = WebIM.conn.getUniqueId();
  var msg = new WebIM.message('cmd', id);
  var currentCallId = info.currentCallId || getState().confr.callId;
  var callerDevId = info.callerDevId || getState().confr.callerDevId;
  var to = info.to || getState().confr.callerIMName;
  msg.set({
    to: to,
    action: 'rtcCall',
    ext: {
      action: 'answerCall',
      result: result,
      // busy/accept/refuse
      callerDevId: callerDevId,
      calleeDevId: WebIM.conn.context.jid.clientResource,
      callId: currentCallId,
      ts: Date.now(),
      msgType: 'rtcCallWithAgora'
    },
    success: function success(id, serverMsgId) {},
    fail: function fail(e) {
      console.warn("callee send answerCall fail", msg);
    }
  });
  WebIM.conn.send(msg.body);
}; // caller

var confirmCallee = function confirmCallee(to, calleeDevId, result) {
  var getState = store.getState,
      dispatch = store.dispatch;
  var id = WebIM.conn.getUniqueId();
  var msg = new WebIM.message('cmd', id);
  var confr = getState().confr;
  var currentCallId = confr.callId;

  if (!confr.calleeDevId && ![2, 3].includes(confr.type)) {
    dispatch(updateConfr({
      to: confr.confrName,
      ext: {
        channelName: confr.channel,
        token: confr.token,
        type: confr.type,
        callerDevId: confr.callerDevId,
        calleeDevId: calleeDevId,
        callId: confr.callId,
        calleeIMName: confr.calleeIMName,
        callerIMName: confr.callerIMName
      }
    }));
  } else if (confr.calleeDevId != calleeDevId && ![2, 3].includes(confr.type)) {
    result = 'refuse';
  }

  msg.set({
    to: to,
    action: 'rtcCall',
    ext: {
      action: 'confirmCallee',
      result: result || 'accept',
      // busy/accept/refuse
      callerDevId: WebIM.conn.context.jid.clientResource,
      calleeDevId: calleeDevId,
      callId: currentCallId,
      ts: Date.now(),
      msgType: 'rtcCallWithAgora'
    },
    success: function success(id, serverMsgId) {
      if (result == 'accept') {
        dispatch(setCallStatus(CALLSTATUS.confirmCallee));
      }
    },
    fail: function fail(e) {
      console.warn("caller send confirmCallee fail", msg);
    }
  });
  WebIM.conn.send(msg.body);
};

export var cancelCall = function cancelCall(to) {
  var getState = store.getState;
  var id = WebIM.conn.getUniqueId();
  var msg = new WebIM.message('cmd', id);
  var callerDevId = getState().confr.callerDevId;
  var user = to || getState().confr.calleeIMName;
  var currentCallId = getState().confr.callId;

  if (!user) {
    console.warn('send cancelCall fail, to is undefined');
    return;
  }

  msg.set({
    to: user,
    action: 'rtcCall',
    ext: {
      action: 'cancelCall',
      callerDevId: callerDevId,
      callId: currentCallId,
      ts: Date.now(),
      msgType: 'rtcCallWithAgora'
    },
    success: function success(id, serverMsgId) {//dispatch(setCallStatus(CALLSTATUS.idle))
    },
    fail: function fail(e) {
      console.warn("caller send canlcel fail", msg);
    }
  });
  WebIM.conn.send(msg.body);
};
export var addListener = function addListener() {
  var getState = store.getState,
      dispatch = store.dispatch;
  WebIM.conn.addEventHandler('message', {
    onTextMessage: function onTextMessage(message) {
      if (message.chatType !== 'singleChat') return;
      var state = getState();
      var conf = state.conf,
          callStatus = state.callStatus;
      var from = message.from,
          to = message.to;

      if (message.ext && message.ext.action === 'invite') {
        if (message.from == WebIM.conn.context.jid.name) {
          return; // msg from other device
        }

        if (message.chatType == 'singleChat') {
          if (callStatus > CALLSTATUS.idle) {
            // busy
            answerCall('busy', {
              currentCallId: message.ext.callId,
              callerDevId: message.ext.callerDevId,
              to: from
            });
          } else {
            message.ext.calleeIMName = message.to;
            message.ext.callerIMName = message.from;
            dispatch(updateConfr(message));
            dispatch(setCallStatus(CALLSTATUS.alerting));
          }
        } else {
          var msgInfo = message.ext;

          if (callStatus > CALLSTATUS.idle) {
            if (msgInfo.callId == conf.callId) {
              dispatch(setCallStatus(CALLSTATUS.alerting));
            } else {
              answerCall('busy', {
                currentCallId: msgInfo.callId,
                callerDevId: msgInfo.callerDevId,
                to: from
              });
            }
          }

          msgInfo.calleeIMName = WebIM.conn.context.jid.name;
          msgInfo.callerIMName = from;
          dispatch(updateConfr({
            from: from,
            to: to,
            ext: msgInfo
          }));
          dispatch(setCallStatus(CALLSTATUS.alerting));
        }
      }
    },
    onCmdMessage: function onCmdMessage(msg) {
      if (msg.action === "rtcCall") {
        var msgInfo = msg.ext;
        var deviceId = '';
        var callerDevId = '';
        var callId = '';
        var callVideo = store.getState();

        switch (msgInfo.action) {
          case "invite":
            return;

            if (msg.from == WebIM.conn.context.jid.name) {
              return; // invite msg send by myself on another device
            }

            var state = getState();
            var conf = state.conf,
                callStatus = state.callStatus;
            var from = msg.from,
                to = msg.to;

            if (callStatus > CALLSTATUS.idle) {
              if (msgInfo.callId == conf.callId) {
                dispatch(setCallStatus(CALLSTATUS.alerting));
              } else {
                answerCall('busy', {
                  currentCallId: msgInfo.callId,
                  callerDevId: msgInfo.callerDevId,
                  to: from
                });
              }
            }

            msgInfo.calleeIMName = WebIM.conn.context.jid.name;
            msgInfo.callerIMName = from;
            dispatch(updateConfr({
              from: from,
              to: to,
              ext: msgInfo
            }));
            dispatch(setCallStatus(CALLSTATUS.alerting));
            break;

          case "alert":
            deviceId = msgInfo.calleeDevId;
            callerDevId = msgInfo.callerDevId;
            callId = msgInfo.callId;
            confirmRing(msg.from, deviceId, callerDevId, callId);
            break;

          case "confirmRing":
            if (msgInfo.calleeDevId != WebIM.conn.context.jid.clientResource) {
              return; // Multi-terminal case A message on the other end (calling a call from another device) caller
            }

            if (!msgInfo.status && callVideo.callStatus < CALLSTATUS.receivedConfirmRing) {
              console.warn('The invitation has expired');
              dispatch(setCallStatus(CALLSTATUS.idle));
              callManager.hangup('invitation has expired'); // WebIM.rtc.timer && clearTimeout(WebIM.rtc.timer)

              return;
            }

            deviceId = msgInfo.calleeDevId;
            dispatch(setCallStatus(CALLSTATUS.receivedConfirmRing)); // answerCall(msg.from, deviceId)
            // WebIM.rtc.timer && clearTimeout(WebIM.rtc.timer)

            break;

          case "answerCall":
            WebIM.rtc.timer && clearTimeout(WebIM.rtc.timer);
            deviceId = msgInfo.calleeDevId;

            if (msgInfo.callerDevId != WebIM.conn.context.jid.clientResource) {
              if (msg.from == WebIM.conn.context.jid.name) {
                if (msgInfo.result === 'accept') {
                  callManager.hangup('accepted on other devices');
                } else if (msgInfo.result === 'refuse') {
                  callManager.hangup('refused on other devices');
                }

                dispatch(setCallStatus(CALLSTATUS.idle));
                WebIM.rtc.timer && clearTimeout(WebIM.rtc.timer);
                return console.warn('processed on other devices');
              }

              return; // 多端情况另一端的消息 （被叫 在其他设备处理）
            }

            if (msgInfo.result !== 'accept') {
              if (msgInfo.result === 'busy') {
                console.warn('target is busy');
              } else if (msgInfo.result === 'refuse') {
                console.error('the target has refused');
              }

              if (![2, 3].includes(callVideo.confr.type)) {
                // 1v1 hang up，multi don't hang up
                confirmCallee(msg.from, deviceId, msgInfo.result);
                callManager.hangup(msgInfo.result);
                dispatch(setCallStatus(CALLSTATUS.idle));
              } else {
                confirmCallee(msg.from, deviceId, 'refuse');
              }
            } else {
              confirmCallee(msg.from, deviceId, msgInfo.result);
            }

            break;

          case "confirmCallee":
            if (msgInfo.calleeDevId != WebIM.conn.context.jid.clientResource && callVideo.callStatus !== 7) {
              callManager.hangup('processed on other devices');
              dispatch(setCallStatus(CALLSTATUS.idle));
              WebIM.rtc.timer && clearTimeout(WebIM.rtc.timer);
              return;
            }

            if (msg.ext.result != 'accept' && callVideo.callStatus !== 7) {
              // Hang up when busy Refuse is received during a call
              if (callVideo.callStatus !== 0) {
                callManager.hangup(msg.ext.result);
                dispatch(setCallStatus(CALLSTATUS.idle));
              }

              WebIM.rtc.timer && clearTimeout(WebIM.rtc.timer);
              return;
            }

            dispatch(setCallStatus(CALLSTATUS.confirmCallee));
            break;

          case "cancelCall":
            if (msg.from == WebIM.conn.context.jid.name) {
              return; // msg from other device
            }

            if (msg.from == callVideo.confr.callerIMName) {
              callManager.hangup('cancel');
              dispatch(setCallStatus(CALLSTATUS.idle));
            }

            break;

          default:
            console.warn("unexpected action " + msg.action);
            break;
        }
      }
    }
  });
};
manager.sendAlerting = sendAlerting;
