// agora-chat-uikit/lib/utils/WebIM.js
"use strict";

exports.__esModule = true;
exports.initIMSDK = exports["default"] = void 0;

var _EasemobChat = _interopRequireDefault(require("./Easemob-chat"));

var _WebIMConfig = _interopRequireDefault(require("./WebIMConfig"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// import websdk from 'easemob-websdk';
// import websdk from 'websdk';
var WebIM = window.WebIM || {};
WebIM.config = _WebIMConfig["default"];

var initIMSDK = function initIMSDK(appkey) {
  WebIM.config = _WebIMConfig["default"];
  WebIM.config.appkey = appkey;
  var options = {
    isMultiLoginSessions: WebIM.config.isMultiLoginSessions,
    isDebug: WebIM.config.isDebug,
    https: WebIM.config.https,
    isAutoLogin: false,
    heartBeatWait: WebIM.config.heartBeatWait,
    autoReconnectNumMax: WebIM.config.autoReconnectNumMax,
    delivery: WebIM.config.delivery,
    appKey: appkey,
    useOwnUploadFun: WebIM.config.useOwnUploadFun,
    deviceId: WebIM.config.deviceId,
    //公有云 isHttpDNS 默认配置为true
    isHttpDNS: WebIM.config.isHttpDNS
  };
  WebIM.conn = new _EasemobChat["default"].connection(options);
};

exports.initIMSDK = initIMSDK;
var _default = WebIM;
exports["default"] = _default;
